### PR TITLE
Issue 15: coding guidelines

### PR DESCRIPTION
**Pull request by LinhWuerzburger**
First attempt to close Issue #15, defining coding guidelines. At least the things which are obvious at the moment.

### CODE REVIEW

**Comment by LinhWuerzburger** [comment]
do you think we should add something like: if you want to contribute, pls open a PR?

**Comment by LinhWuerzburger** [comment]
I changed the name from contributing to coding guidelines because it only has information on the coding guidelines and nothing more. a guide for contributing is in the wiki. the question is: do we want to link to the wiki or not? Or should we make it the other way around and link in the wiki to a markdown file?

**Review by ChristianVonMach — CHANGES_REQUESTED** [review]
There are some other points in the Googles style guide, that needs to bee addressed. I don't think we should handle all at onces. So this are the thirst few points.

File: CODING_GUIDELINES.md [hunk 1]
```diff
@@ -0,0 +1,52 @@
+## Contributing Guidelines
+
+The coding guidelines of ARTSS are mainly based on the [google coding guidelines](https://google.github.io/styleguide/cppguide.html). Below there are some rules which are not defined in the google standard or we want to especially emphasise them in order to achieve the most consistent code possible.
```
[the remark this answers is not in the document]
Comment (LinhWuerzburger) [reply CODING_GUIDELINES.md]: We don't use C++20 yet see #105 
Regarding the second point, I don't know what you mean with non-standard extension?

[the remark this answers is not in the document]
Comment (LinhWuerzburger) [reply CODING_GUIDELINES.md]: done in 3daa8cf

[the remark this answers is not in the document]
Comment (LinhWuerzburger) [reply CODING_GUIDELINES.md]: If I understand it correctly, you are planning to put every class into a namespace and not only the old static classes? And accordingly to add the rule to put everything in a namespace?

[the remark this answers is not in the document]
Comment (LinhWuerzburger) [reply CODING_GUIDELINES.md]: Can you formulate a passage regarding these rules?

[the remark this answers is not in the document]
Comment (LinhWuerzburger) [reply CODING_GUIDELINES.md]: if adjusted how do you plan to adjust it or do you want to drop the rule?

[the remark this answers is not in the document]
Comment (LinhWuerzburger) [reply CODING_GUIDELINES.md]: I see the benefit in trailing return type syntax only for Lamba expressions. Therefore I propose to adjust the rule to 'for Lambda expressions the trailing return type syntax is used'. Or something similar.

[the remark this answers is not in the document]
Comment (LinhWuerzburger) [reply CODING_GUIDELINES.md]: Is the trailing return type syntax something which can be used for our expressions in GlobalMacrosTypes?

**Comment by LinhWuerzburger** [comment]
I changed from relying on the google standard to our very own rules which are based on the google standard and in addition, I added that it is NOT necessary to know the google standard. Therefore we have to add every rule from the google standard which we want to use. If the rule is a large passage we can refer to the google standard for more detailed information.

**Comment by LinhWuerzburger** [comment]
And I don't want to add the rule to not exceed 80 characters per line. This is just way too short. But we should add a section to line breaking. I don't mind to change the limit or to change the whole rule and to say that there won't be a line break at all. But I am not going with 80 characters.

**Comment by LinhWuerzburger** [comment]
> And I don't want to add the rule to not exceed 80 characters per line. This is just way too short. But we should add a section to line breaking. I don't mind to change the limit or to change the whole rule and to say that there won't be a line break at all. But I am not going with 80 characters.

We could agree on a 100 character hard break while trying to take the 80 character line.

File: CODING_GUIDELINES.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (ChristianVonMach) [reply CODING_GUIDELINES.md]: yeas every class should be in a namespace but of course not it's own. I my opinion we should use namespaces to represent the directories in our code like solver or pressure.

[the remark this answers is not in the document]
Comment (ChristianVonMach) [reply CODING_GUIDELINES.md]: "ARTSS currently uses some static and global variables, but do not add new ones except for a exceptional good reason."

[the remark this answers is not in the document]
Comment (ChristianVonMach) [reply CODING_GUIDELINES.md]: trailing return types doesn't work on macros. we still have to relay on our compiler to deduce the correct type.

[the remark this answers is not in the document]
Comment (ChristianVonMach) [reply CODING_GUIDELINES.md]: I think if we go with the rule of 3-5-0 we are going to be okay.

[the remark this answers is not in the document]
Comment (LinhWuerzburger) [reply CODING_GUIDELINES.md]: fine by me. Could you open an issue regarding this discussion? I will think of a text for the guidelines.

[the remark this answers is not in the document]
Comment (LinhWuerzburger) [reply CODING_GUIDELINES.md]: [3-5-0](https://google.github.io/styleguide/cppguide.html#Structs_vs._Tuples) doesn't seem the rule you are speaking of

**Review by ChristianVonMach — APPROVED** [review]

**Pull request merged**